### PR TITLE
Implement option to disable cache rebuild if it already exists in storage

### LINKED ISSUE #83: Cache is always rebuilding
The cache is currently always being rebuilt for me—I think we should compute the `cache_key` in the rebuild process, and if it has changed then we rebuild.

```
cache_key: {{ checksum "yarn.lock" }}
```

That's my `cache_key`—if the lock file hasn't changed (hence the key is the same), there's no need to re-upload `node_modules` entirely.

If implemented, should also check if the cache for the given key exists. Otherwise, it's the first time we're generating it.

Thoughts?

### PR DESCRIPTION
**Pull request by lafriks**
### Description

Adds `override` option that is true by default to preserve backward compatibility

Fixes #83

### Checklist

- [x] Read the **CONTRIBUTING** document. (It's checked since you are already here.)
- [x] Read the [**CODE OF CONDUCT**](CODE_OF_CONDUCT.md) document.
- [x] Add tests to cover changes.
- [x] Ensure your code follows the code style of this project.
- [ ] Ensure CI and all other PR checks are green OR
    - [x] Code compiles correctly.
    - [x] Created tests which fail without the change (if possible).
    - [x] All new and existing tests passed.
- [x] Add your changes to `Unreleased` section of [CHANGELOG](CHANGELOG.md).
- [x] Improve and update the [README](README.md) (if necessary).
- [x] Ensure [documentation](./DOCS.md) is up-to-date. The same file will be updated in [plugin index](https://github.com/drone/drone-plugin-index/blob/master/content/meltwater/drone-cache/index.md) when your PR is accepted, so it will be available for end-users at http://plugins.drone.io.

### CODE REVIEW

**Comment by lafriks** [comment]
I don't think test failure is related to this PR

**Comment by lafriks** [comment]
Ok, might be just tests are not stable, after force push now they succeeded

**Comment by kakkoyun** [comment]
@lafriks CI failures are unrelated, I've opened an issue about it https://github.com/meltwater/drone-cache/issues/104.

Checking the rest of the PR now.

File: storage/backend/s3/s3.go [hunk 1]
```diff
@@ -114,3 +115,22 @@ func (b *Backend) Put(ctx context.Context, p string, r io.Reader) error {
...
+	out, err := b.client.HeadObjectWithContext(ctx, in)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == s3.ErrCodeNoSuchKey || awsErr.Code() == "NotFound" {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("head the object, %w", err)
+	}
+
+	return *out.ETag != "", nil
```
Comment (kakkoyun) [inline storage/backend/s3/s3.go:135]: Maybe we should add a small comment about the emptiness of `out.ETag`.

File: storage/backend/gcs/gcs.go [hunk 2]
```diff
@@ -140,6 +140,41 @@ func (b *Backend) Put(ctx context.Context, p string, r io.Reader) error {
 	}
 }
 
+// Exists checks if object already exists.
+func (b *Backend) Exists(ctx context.Context, p string) (bool, error) {
+	errCh := make(chan error)
```
Comment (kakkoyun) [inline storage/backend/gcs/gcs.go:145]: Although I'm not completely sure if it'd improve anything, we can introduce an inline result struct `result{bool, err}` and use only a single channel to collect results. What do you think?

File: cache/rebuilder.go [hunk 3]
```diff
@@ -58,6 +59,17 @@ func (r rebuilder) Rebuild(srcs []string) error {
 
 		dst := filepath.Join(namespace, key, src)
 
+		// If no override is set and object already exists in storage, skip it
```
Comment (kakkoyun) [inline cache/rebuilder.go:62]:
```suggestion
		// If no override is set and object already exists in storage, skip it.
```

File: cache/option.go [hunk 4]
```diff
@@ -31,3 +32,10 @@ func WithFallbackGenerator(g key.Generator) Option {
 		o.fallbackGenerator = g
 	})
 }
+
+// WithOverride sets object should be overriten even if it already exists.
+func WithOverride() Option {
```
Comment (kakkoyun) [inline cache/option.go:37]: This should take the value and set it. 
If nothing, it gives the ability to change behaviour when we test it.

File: internal/plugin/plugin.go [hunk 5]
```diff
@@ -95,6 +95,10 @@ func (p *Plugin) Exec() error {
 		options = append(options, cache.WithFallbackGenerator(keygen.NewStatic(p.Metadata.Commit.Branch)))
 	}
 
+	if p.Config.Override {
+		options = append(options, cache.WithOverride())
```
Comment (kakkoyun) [inline internal/plugin/plugin.go:99]: As I commented above, just pass to value to options.
Let's make changing override value to false explicit.

**Comment by lafriks** [comment]
@kakkoyun done

**Comment by kakkoyun** [comment]
This will be in the next stable release `v1.1.0`, it shouldn't take long. I'm planning to wrap things up and tag a release candidate this weekend. Until then you can pick up a master build from https://hub.docker.com/r/meltwater/drone-cache/tags?page=1&ordering=last_updated to use in your pipeline.

**Comment by lafriks** [comment]
Could you restart drone master build task? :)

**Comment by kakkoyun** [comment]
CI is killing me 😓

**Pull request merged**